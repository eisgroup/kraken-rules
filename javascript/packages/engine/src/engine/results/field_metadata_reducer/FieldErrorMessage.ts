/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { PayloadResultType } from 'kraken-engine-api'
import { Payloads } from 'kraken-model'

/**
 * Contains validation error message from result of validation rule evaluation
 */
export class FieldErrorMessage {
    constructor(
        public readonly ruleName: string,
        public readonly message: string,
        public readonly errorCode: string,
        public readonly payloadResultType: PayloadResultType,
        public readonly severity?: Payloads.Validation.ValidationSeverity,
    ) {}
}
