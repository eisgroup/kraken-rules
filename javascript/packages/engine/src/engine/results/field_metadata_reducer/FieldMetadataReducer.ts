/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
    RuleEvaluationResults,
    EntryPointResult,
    RuleOverride,
    ContextFieldInfo,
    ErrorMessage,
    FieldMetadata,
    FieldMetadataResult,
} from 'kraken-engine-api'

import { EntryPointReducer } from '../Reducer'
import { Localization } from '../Localization'
import ValidationRuleEvaluationResult = RuleEvaluationResults.ValidationRuleEvaluationResult
import { payloadResultTypeChecker } from '../PayloadResultTypeChecker'

/**
 * Processes rule evaluation information from {@link EntryPointResult}, reducing it to list of
 * metadata entries for each distinct field
 */
export class FieldMetadataReducer implements EntryPointReducer<Record<string, FieldMetadata>> {
    readonly #isRuleOverridden: RuleOverride.IsRuleOverridden
    readonly #errorMessageFrom: (rer: ValidationRuleEvaluationResult, contextInfo: ContextFieldInfo) => ErrorMessage

    constructor(
        isRuleOverridden?: RuleOverride.IsRuleOverridden,
        validationMessageProvider?: Localization.ValidationMessageProvider,
    ) {
        this.#isRuleOverridden = isRuleOverridden ? isRuleOverridden : () => false

        if (validationMessageProvider) {
            this.#errorMessageFrom = rer => Localization.resolveMessage(rer.payloadResult, validationMessageProvider)
        } else {
            this.#errorMessageFrom = rer =>
                Localization.resolveMessage(rer.payloadResult, Localization.defaultValidationMessages)
        }
    }

    /**
     * @override
     */
    reduce(results: EntryPointResult): Record<string, FieldMetadata> {
        const fieldMetadataResults: Record<string, FieldMetadata> = {}
        const fieldResults = results.getFieldResults()
        for (const id in fieldResults) {
            if (Object.prototype.hasOwnProperty.call(fieldResults, id)) {
                const fieldResult = fieldResults[id]
                let isDisabled
                let isHidden
                const validationResults: FieldMetadataResult[] = []
                for (const ruleResult of fieldResult.ruleResults) {
                    if (!RuleEvaluationResults.isNotApplicable(ruleResult)) {
                        const payloadResult = ruleResult.payloadResult
                        // resolve accessibility
                        if (!isDisabled && payloadResultTypeChecker.isAccessibility(payloadResult)) {
                            isDisabled = !payloadResult.accessible
                        }
                        // resolve visibility
                        if (!isHidden && payloadResultTypeChecker.isVisibility(payloadResult)) {
                            isHidden = !payloadResult.visible
                        }
                        // resolve validation
                        if (RuleEvaluationResults.isValidation(ruleResult)) {
                            const error = this.#errorMessageFrom(ruleResult, fieldResult.contextFieldInfo)
                            const isOverridden =
                                RuleEvaluationResults.overrideInfoTypeChecker.isOverrideApplicable(
                                    ruleResult.overrideInfo,
                                ) && this.#isRuleOverridden(ruleResult.overrideInfo.overrideContext)
                            validationResults.push({
                                rawTemplateVariables: ruleResult.payloadResult.message?.rawTemplateVariables ?? [],
                                ruleName: ruleResult.ruleInfo.ruleName,
                                errorCode: error.errorCode,
                                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                                errorMessage: error.errorMessage!,
                                templateVariables: error.templateVariables,
                                payloadResult,
                                severity: ruleResult.payloadResult.validationSeverity,
                                isFailed:
                                    ruleResult.payloadResult.success !== undefined
                                        ? !ruleResult.payloadResult.success
                                        : undefined,
                                isOverridable: ruleResult.overrideInfo.overridable,
                                isOverridden,
                            })
                        }
                    }
                }

                // register results
                fieldMetadataResults[id] = {
                    id,
                    info: {
                        getContextInstanceId: () => fieldResult.contextFieldInfo.contextId,
                        getContextName: () => fieldResult.contextFieldInfo.contextName,
                    },
                    resolvedTargetPath: fieldResult.contextFieldInfo.fieldPath,
                    isDisabled,
                    isHidden,
                    ruleResults: validationResults,
                    fieldType: fieldResult.contextFieldInfo.fieldType,
                }
            }
        }
        return fieldMetadataResults
    }
}
