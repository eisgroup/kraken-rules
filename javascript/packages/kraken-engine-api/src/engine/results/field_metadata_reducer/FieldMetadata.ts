/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Payloads } from 'kraken-model'
import { PayloadResult } from '../PayloadResult'
import { ContextInstanceInfo } from '../../contexts/info/ContextInstanceInfo'

export interface FieldMetadataResult {
    ruleName: string
    errorCode: string
    errorMessage: string
    templateVariables: string[]
    payloadResult: PayloadResult
    isOverridable: boolean
    isOverridden: boolean
    severity?: Payloads.Validation.ValidationSeverity
    /**
     * if field is undefined, then rule evaluation was ignored
     */
    isFailed?: boolean
}

/**
 * Metadata that will be returned in map {field:metadata}. It contains all info after evaluating rules on field.
 */
export interface FieldMetadata {
    id: string
    resolvedTargetPath: string
    info: ContextInstanceInfo
    isHidden?: boolean
    isDisabled?: boolean
    ruleResults: FieldMetadataResult[]
    fieldType: string
}
