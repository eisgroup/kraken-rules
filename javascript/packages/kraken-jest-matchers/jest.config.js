/*
 * Copyright Â© 2022 EIS Group and/or one of its affiliates. All rights reserved. Unpublished work under U.S. copyright laws.
 * CONFIDENTIAL AND TRADE SECRET INFORMATION. No portion of this work may be copied, distributed, modified, or incorporated into any other
 * media without EIS Group prior written consent.
 */
module.exports = {
    moduleFileExtensions: ['ts', 'tsx', 'js'],
    transform: {
        '^.+\\.ts$': ['@swc/jest'],
    },
    testMatch: ['**/?(*.)test.ts'],
    verbose: true,
    cacheDirectory: './target/tmp/',
    coverageDirectory: './target/coverage',
    coverageThreshold: {
        global: {
            branches: 60,
            functions: 80,
            lines: 80,
            statements: 80,
        },
    },
    collectCoverageFrom: ['**/*[^d\\.].ts', '!**/node_modules/**', '!**/target/**', '!**/__tests__/**', '!**/index.ts'],
    testEnvironment: 'node',
    setupFilesAfterEnv: [],
}
