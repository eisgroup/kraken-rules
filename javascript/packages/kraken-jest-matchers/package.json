{
    "name": "kraken-jest-matchers",
    "version": "1.47.2-stage.0",
    "description": "Additional Jest matchers to use in Kraken Tests",
    "main": "target/dist/js/src/index.js",
    "types": "target/dist/definitions/src/index.d.ts",
    "author": "Tomas Dapkunas <tdapkunas@eisgroup.com> (https://eisgroup.com)",
    "license": "Apache-2.0",
    "publishConfig": {
        "registry": "https://genesis-npm-release.exigengroup.com/repository/genesis-npm-release/"
    },
    "typedocMain": "src/index.ts",
    "dependencies": {
        "declarative-js": "3.8.15",
        "kraken-engine-api": "1.47.2-stage.0",
        "kraken-model": "1.47.2-stage.0"
    },
    "devDependencies": {
        "@types/jest": "26.0.22",
        "@types/lodash": "4.14.116",
        "@types/node": "13.9.0",
        "child_process": "1.0.2",
        "fast-xml-parser": "^3.19.0",
        "jest": "26.6.3",
        "jsdom": "11.3.0",
        "jsdom-global": "3.0.2",
        "prettier": "1.14.3",
        "rimraf": "3.0.2",
        "ts-jest": "26.5.0",
        "ts-node": "4.1.0",
        "typescript": "4.2.4"
    },
    "peerDependencies": {
        "jest-matcher-utils": "*",
        "jest-snapshot": "*"
    },
    "scripts": {
        "lint": "eslint src",
        "test": "jest -c jest.config.js",
        "test:coverage": "jest -c jest.config.js --coverage",
        "compile": "tsc --pretty",
        "postcompile": "yarn transpile",
        "transpile": "echo 'no transpile'",
        "del:compiled": "rimraf target",
        "build": "yarn run compile && yarn run lint && yarn run test:coverage",
        "prepare": "yarn run del:compiled && yarn run compile && yarn run lint --fix && yarn run test:coverage",
        "release": "yarn publish"
    },
    "files": [
        "target/dist/definitions",
        "target/dist/js"
    ],
    "gitHead": "e0a70c3ae821b768a1706c1afff4e77bfa5886e5"
}