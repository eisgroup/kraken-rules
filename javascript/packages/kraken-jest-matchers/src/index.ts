/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { matchers as dateTimeMatchers } from './dateTimeMatchers'
import { matchers as ruleEvalResultMatchers } from './ruleEvalResultMatchers'
import { matchers as entryPointResultMatchers } from './entrypointResultMatchers'

export const matchers = {
    ...dateTimeMatchers,
    ...ruleEvalResultMatchers,
    ...entryPointResultMatchers,
}

export { ExpectedResult } from './entrypointResultMatchers'
export { MatchableResults } from './entrypointResultMatchers'

expect.extend(matchers)
