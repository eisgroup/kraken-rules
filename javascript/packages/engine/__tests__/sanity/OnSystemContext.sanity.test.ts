/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { sanityMocks } from './_AutoPolicyObject.mocks'
import { sanityEngine } from './_SanityEngine'
import { FieldMetadataReducer } from '../../src/engine/results/field_metadata_reducer/FieldMetadataReducer'
import { MatchableResults } from 'kraken-jest-matchers'

describe('Engine Sanity On System Context Test', () => {
    const { empty } = sanityMocks
    it("should execute 'on system context' entrypoint", () => {
        const policy = empty()
        policy.backupSystemDate = { date: '2020-02-02T02:02:02' }

        const results = sanityEngine.evaluate(policy, 'on system context')

        const fieldResults = new FieldMetadataReducer().reduce(results)
        const matchableResults = { entryPointResults: results, fieldResults } as MatchableResults

        expect(results).k_toMatchResultsStats({ total: 1 })
        expect(results).k_toHaveExpressionsFailures(0)
        expect(matchableResults).k_toMatchResultsSnapshots()
        expect(policy.systemDate).toBe(policy.backupSystemDate)
    })
})
