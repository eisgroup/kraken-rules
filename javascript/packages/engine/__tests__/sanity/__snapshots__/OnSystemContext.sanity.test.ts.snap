// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine Sanity On System Context Test should execute 'on system context' entrypoint 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:systemDate": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "systemDate",
          "fieldPath": "systemDate",
          "fieldType": "SysDate",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "systemDate",
                  "contextId": "0",
                  "contextName": "Policy",
                  "newValue": Object {
                    "date": "2020-02-02T02:02:02",
                  },
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "DEFAULT",
              "ruleName": "on system context",
              "targetPath": "systemDate",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:0:systemDate": Object {
      "fieldType": "SysDate",
      "id": "Policy:0:systemDate",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "systemDate",
      "ruleResults": Array [],
    },
  },
}
`;
